lib/langchain/assistants/assistant.rb

```diff
@@ -79,54 +79,10 @@ def run(auto_tool_execution: false)
         return
       end
 
-      running = true
-
-      while running
-        # TODO: I think we need to look at all messages and not just the last one.
-        last_message = thread.messages.last
-
-        if last_message.system?
-          # Do nothing
-          running = false
-        elsif last_message.llm?
-          if last_message.tool_calls.any?
-            if auto_tool_execution
-              run_tools(last_message.tool_calls)
-            else
-              # Maybe log and tell the user that there's outstanding tool calls?
-              running = false
-            end
-          else
-            # Last message was from the assistant without any tools calls.
-            # Do nothing
-            running = false
-          end
-        elsif last_message.user?
-          # Run it!
-          response = chat_with_llm
-
-          if response.tool_calls.any?
-            # Re-run the while(running) loop to process the tool calls
-            running = true
-            add_message(role: response.role, tool_calls: response.tool_calls)
-          elsif response.chat_completion
-            # Stop the while(running) loop and add the assistant's response to the thread
-            running = false
-            add_message(role: response.role, content: response.chat_completion)
-          end
-        elsif last_message.tool?
-          # Run it!
-          response = chat_with_llm
-          running = true
-
-          if response.tool_calls.any?
-            add_message(role: response.role, tool_calls: response.tool_calls)
-          elsif response.chat_completion
-            add_message(role: response.role, content: response.chat_completion)
-          end
-        end
-      end
+      state = :in_progress
+      state = handle_state(state, auto_tool_execution) until run_finished?(state)
 
+      # TODO: Should we return the final state along with the messages?
```

Comment:
I think it's okay to skip. But I think the developer should be able to access it with:
```ruby
assistant.state
#=> :in_progress
```
What do you think?

Reply:
Yes, adding a `state` attribute to the assistant could be a valid solution without modifying the return value of the method.

A note: In synchronous execution, `assistant.state` will always return the final state (`:completed`, `:failed`, or `:expired`) because during the `run` method, you cannot access the attribute. Therefore, you will never see `:in_progress` or `:running_tools`. Correct?

Reply:
Wouldn't you see the `:running_tools` (`:requires_action`)? For example:
```ruby
irb(main):019> a.add_message content:"Latests news in politics?"
=>
[#<Langchain::Messages::OpenAIMessage:0x000000011f1de708 @content="You are a news reporter", @role="system", @tool_call_id=nil, @tool_calls=[]>,
 #<Langchain::Messages::OpenAIMessage:0x00000001202b2ae8 @content="Latests news in politics?", @role="user", @tool_call_id=nil, @tool_calls=[]>]
irb(main):020> a.run auto_tool_execution: false
I, [2024-06-25T13:10:51.775951 #77146]  INFO -- : [Langchain.rb] [Langchain::Assistant]: Sending a call to Langchain::LLM::OpenAI
=>
[#<Langchain::Messages::OpenAIMessage:0x000000011f1de708 @content="You are a news reporter", @role="system", @tool_call_id=nil, @tool_calls=[]>,
 #<Langchain::Messages::OpenAIMessage:0x00000001202b2ae8 @content="Latests news in politics?", @role="user", @tool_call_id=nil, @tool_calls=[]>,
 #<Langchain::Messages::OpenAIMessage:0x000000012011a370
  @content="",
  @role="assistant",
  @tool_call_id=nil,
  @tool_calls=
   [{"id"=>"call_MnJgY7DHwfPT1yPXRM1KRF1H",
     "type"=>"function",
     "function"=>{"name"=>"news_retriever__get_top_headlines", "arguments"=>"{\"category\":\"general\",\"q\":\"politics\",\"page_size\":5}"}}]>]
irb(main):021>
```
At this point the `state` would be `:running_tools` (or `:requires_action`), correct?

Reply:
No, as you can see in `handle_llm_message`, if auto tool execution is disabled, the return value is currently `:completed`. This is necessary to exit the loop (see `run_finished?`).

```ruby
if last_message.tool_calls.any?
  auto_tool_execution ? :running_tools : :completed
else
```

As written in the TODO notes, I was thinking to implement a new state (like `:requires_manual_action`). Or we should find a way to exit while keeping the state `requires_action`. What do you think?

Reply:
@dghirardo My gut feeling is to simplify, so less states. Do you agree? It's much easier to add new states than to try to simplify it later.

Reply:
One last question. When the assistant is initialized for the first time, what will `state` be set to? Nil?

Reply:
Good question. Either "new" or "initialized" or "ready". What do you think?